File: go/ex4_1.go
package main

type List[T any] interface {
    get(idx int) T
}

type ArrayList[T any] struct { t T }
func (l ArrayList[T]) get(idx int) T {
    return l.t
}

type BitVector struct {}
func (l BitVector) get(idx int) bool {
    return true
}

func main() {
    var l1 List[string] = ArrayList[string]{""}
    var l2 List[bool] = ArrayList[bool]{true}
    var l3 List[bool] = BitVector{}
    _ = l1
    _ = l2
    _ = l3
}

File: typescript/ex4_1.ts
abstract class List<T> {
    abstract get(idx: number): T;
}
class ArrayList<T> {
    v: T;
    constructor(v: T) { this.v = v; }
    get(idx: number): T { return this.v; }
}
class BitVector {
    get(idx: number): boolean { return true; }
}

let l1: List<string> = new ArrayList<string>("");
let l2: List<boolean> = new ArrayList<boolean>(true);
let l3: List<boolean> = new BitVector();
